### PR TITLE
error returning epoch hooks

### PR DESCRIPTION
**Pull request by puneet2019**
Closes: #XXX
A part of #2520 only implemented for epochs module. 
If it looks good, can go ahead and implement for other modules as well. 
would wait for #2519 to be merged before. (rebase, etc. then open up)

## What is the purpose of the change

- Hooks errors can now be handled with [`ApplyFuncIfNoError`](https://github.com/osmosis-labs/osmosis/blob/main/osmoutils/cache_ctx.go#L16) 
- This implementation allows `hooks interface` functions to return an error.
- If functions return an error any state changes will not be written to the state.
- Currently it is only done for [`EpochHooks interface`](https://github.com/osmosis-labs/osmosis/blob/main/x/epochs/types/hooks.go#L9)

## Brief Changelog

- Epoch Hooks now allow returning of error.
- All modules implementing EpochHooks also return error in place of panic.

## Testing and Verifying
- Updated and fixed test cases :::NOTE: unable to verify errorMessage.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? - no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no) yes
  - How is the feature or change documented? - not applicable

### CODE REVIEW

File: x/epochs/keeper/hooks.go [hunk 1]
```diff
@@ -5,11 +5,11 @@ import (
...
 // AfterEpochEnd gets called at the end of the epoch, end of epoch is the timestamp of first block produced after epoch duration.
-func (k Keeper) AfterEpochEnd(ctx sdk.Context, identifier string, epochNumber int64) {
-	k.hooks.AfterEpochEnd(ctx, identifier, epochNumber)
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, identifier string, epochNumber int64) error {
+	return k.hooks.AfterEpochEnd(ctx, identifier, epochNumber)
 }
 
 // BeforeEpochStart new epoch is next block of epoch end block
-func (k Keeper) BeforeEpochStart(ctx sdk.Context, identifier string, epochNumber int64) {
-	k.hooks.BeforeEpochStart(ctx, identifier, epochNumber)
```
Comment (ValarDragon) [inline x/epochs/keeper/hooks.go:14]: Can we ignore the error in here, instead of in abci.go?

File: x/mint/keeper/hooks_test.go [hunk 2]
```diff
@@ -395,7 +395,10 @@ func (suite *KeeperTestSuite) TestAfterEpochEnd() {
 
 			osmoassert.ConditionalPanic(suite.T(), tc.expectedPanic, func() {
```
Comment (ValarDragon) [inline x/mint/keeper/hooks_test.go:396]: We should change the variable here to `expectedError`, and then just do an if statement for checking if error vs checking no error

**Review by ValarDragon — APPROVED** [review]
This LGTM, thanks for doing this!

**Review by p0mvn — APPROVED** [review]

File: x/epochs/keeper/hooks.go (hunk 1 above)
Comment (puneet2019) [reply x/epochs/keeper/hooks.go]: done!

File: x/mint/keeper/hooks_test.go (hunk 2 above)
Comment (puneet2019) [reply x/mint/keeper/hooks_test.go]: done!

**Review by p0mvn — APPROVED** [review]
All suggestions have been applied. Should be ready for merge. Thanks @puneet2019

**Pull request merged**